libs/teams/teams-chat-workflow-spring-boot-starter/src/main/java/org/finos/springbot/teams/handlers/retry/InMemoryRetryingActivityHandler.java

```diff
@@ -0,0 +1,37 @@
+package org.finos.springbot.teams.handlers.retry;
+
+import java.time.LocalDateTime;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import org.finos.springbot.teams.conversations.TeamsConversations;
+
+public class InMemoryRetryingActivityHandler extends AbstractRetryingActivityHandler  {
+
+	private Queue<MessageRetry> queue = new ConcurrentLinkedQueue<>();
```

Comment:
Have you looked at this?  https://stackoverflow.com/questions/40485398/retry-logic-with-completablefuture

Reply:
I have already tried this, it was not working. As per our requirement if exception occurred we need to wait for Retry-After then try retry.

Reply:
should not be too hard to add a delay on the .exceptionally() line, if that is the problem

Reply:
No our requirement is to wait for Retry-Afer time, but we can't add the sleep or wait as it will wait the thread. we have implemented similar only inserted data into queue. and added cron job to resent the data.

Reply:
but we have this code to get the retry after time:

```
public long getRetryAfter(CompletionException e) {
		ErrorResponseException ere = (ErrorResponseException) ((CompletionException) e).getCause();
		retrofit2.Response<ResponseBody> response = ere.response();
		String retryAfter = response.headers().get("Retry-After");

		long retryAfterInt = 1;// initiate to 1 sec
		if (StringUtils.isNumeric(retryAfter)) {
			retryAfterInt = Long.parseLong(retryAfter);
		}

		return retryAfterInt;
	}
```


So couldn't you just have some code like this:

```
@Override
	public CompletableFuture<ResourceResponse> handleActivity(Activity activity, TeamsAddressable to) {
		CompletableFuture<ResourceResponse> f= tc.handleActivity(activity, to);
	    for(int i=0; i<teamsRetryCount; i++) {
	        f=f.thenApply(CompletableFuture::completedFuture)
	           .exceptionally(t -> {
	        	   if (isTooManyRequest(t)) {
	        		   long ra = getRetryAfter((CompletionException) t);
	        		   Thread.sleep(ra);
	        		   return  tc.handleActivity(activity, to);
	        	   } else {
	        		   return CompletableFuture.failedFuture(t);
	        	   }
	        	   
	           })
	           .thenCompose(Function.identity());
	    }
	    return f;
	}

```

Apologies if you already tried this - I'm also trying to get my head around `CompletableFuture`s!

Reply:
argh you're right!! 😦 

How about this: https://github.com/deutschebank/spring-bot/pull/1

I actually wrote a test here - well, changed your `InMemoryRetryingActivityHandlerTest` to see if this actually worked.

Test passes but maybe you have more to add..?

Reply:
Hi Rob,

Testcase only pass, if we have `for (int i = 0; i < teamsRetryCount; i++) {` for loop in code. But in our case, if exception occurred we simply insert that record in in-memory queue. 
Here we have to check the how to write testcase around schedule also.

Reply:
Hi, we can wait until Wednesday to talk about this or we can maybe have a call a bit sooner?